### ISSUE TITLE
runtime: TestStructHash, TestArrayHash etc failing on plan9-arm

### ISSUE THREAD

**Issue by millerresearch** [golang/go#45090]
After [CL280372](https://go-review.googlesource.com/c/go/+/280372), hash tests are failing on plan9-arm:
```
--- FAIL: TestSmhasherAppendedZeros (0.00s)
    hash_test.go:161: unexpected number of collisions: got=92 mean=0.000008 stddev=0.002821 threshold=50.423229
--- FAIL: TestSmhasherSmallKeys (0.03s)
    hash_test.go:161: unexpected number of collisions: got=256 mean=0.503906 stddev=0.709864 threshold=156.983435
--- FAIL: TestSmhasherZeros (0.00s)
    hash_test.go:161: unexpected number of collisions: got=127 mean=0.000122 stddev=0.011054 threshold=51.658213
--- FAIL: TestArrayHash (0.00s)
    hash_test.go:703: too many allocs 8.000000 - hash not balanced
--- FAIL: TestStructHash (0.00s)
    hash_test.go:759: too many allocs 8.000000 - hash not balanced
FAIL
FAIL	runtime	76.325s
```
Also failing in hash/maphash:
```
--- FAIL: TestSmhasherAppendedZeros (0.00s)
    smhasher_test.go:99: unexpected number of collisions: got=70 mean=0.000000 stddev=0.000000
--- FAIL: TestSmhasherZeros (0.00s)
    smhasher_test.go:99: unexpected number of collisions: got=864 mean=0.000000 stddev=0.000000
FAIL
FAIL	hash/maphash	0.187s
```

**Comment by randall77** [comment]
Weird that this is happening on this 32-bit platform and not others.

@mengzhuo

**Comment by mengzhuo** [comment]
This will affect on plan9/386 too.
Plan9 ONLY use extendRandom for initial hashkey generate while other platforms read OS random into hashkey first.

https://github.com/golang/go/blob/f5e6d3e879f487066d1a05b8000a7187247558f7/src/runtime/os_plan9.go#L326-L329

This will always had 0 in the first byte. We have three options for this issue:

1. Get random data from OS on plan9 ( I prefer this one but I don't have any plan9 to check.   @millerresearch Can you help us?
2. Xor constant primes in the mix32 (slower and it will got failed if data is identical to hashkey: 2^32 -1 )
3. Drop wyhash and revert to cityhash

**Comment by tmthrgd** [comment]
[CL 6386](https://golang.org/cl/6386) is relevant. Plan9 used to read from /dev/random before that CL, but that was intentionally removed because it was "slow and [...] may block". Perhaps things have changed since 2015 though.

**Comment by mengzhuo** [comment]
@tmthrgd 

CL 6386 is right.
It much more slower and I think maybe we can put const primes into hashkey first for plan9.

```
read /dev/random
ok  	archive/tar	3.241s
ok  	archive/zip	4.145s
ok  	bufio	3.142s
ok  	bytes	16.037s
ok  	compress/bzip2	2.547s
ok  	compress/flate	3.316s
````

```
without /dev/random
ok  	archive/tar	0.257s
ok  	archive/zip	1.246s
ok  	bufio	0.286s
ok  	bytes	12.756s
```

**Comment by mengzhuo** [comment]
@wangyi-fudan Any idea how can we initial hashkey without any entropy input?

**Comment by millerresearch** [comment]
@0intro, is the GCE-specific slowness described in #10028 still the case? Maybe it's time to investigate that.
What virtualisation technology does GCE use? Can I replicate it at home for debugging or do I need to set up a GCE account and try to do it remotely?

**Comment by 0intro** [comment]
I don't think that's specific to GCE. `/dev/random` has always been slow and Plan 9 and can block (like `/dev/random` on Linux).

At some point, I wrote [randfs](http://9legacy.org/9legacy/patch/randfs.diff), which replaces `/dev/random` by a, much faster, pseudo-random generator (similar to `/dev/urandom` on Linux). It solved the issue, but we finally changed `getRandomData`,
because we didn't really need true random number just to improve the distribution of hash values.

If we're going to use the Plan 9's `/dev/random`, we will have to reintroduce [randfs](http://9legacy.org/9legacy/patch/randfs.diff) as a dependency to run Go on Plan 9, which wouldn't not be very desirable.

I'd prefer options 2. or 3.
